engines/router/missionctl/handlers/batch_http_handler.go

```diff
@@ -0,0 +1,124 @@
+	httpHandler
+}
+
+type batchResponse struct {
+	StatusCode int         `json:"code"`
+	ErrorMsg   string      `json:"error,omitempty"`
+	Data       interface{} `json:"data,omitempty"`
+}
+
+type batchResult struct {
+	BatchResult []batchResponse `json:"batch_result"`
+}
+
+// NewBatchHTTPHandler creates an instance of the Mission Control's prediction request handler
+func NewBatchHTTPHandler(mc missionctl.MissionControl) http.Handler {
+	return &batchHTTPHandler{httpHandler{mc}}
+}
+
+func (h *batchHTTPHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	var httpErr *errors.HTTPError
+	h.measureRequestDuration(httpErr)
+
+	// Create context from the request context
+	ctx := turingctx.NewTuringContext(req.Context())
+	// Create context logger
+	ctxLogger := log.WithContext(ctx)
+	defer func() {
+		_ = ctxLogger.Sync()
+	}()
+
+	// Get the unique turing request id from the context
+	turingReqID, err := turingctx.GetRequestID(ctx)
+	if err != nil {
+		ctxLogger.Errorf("Could not retrieve Turing Request ID from context: %v",
+			err.Error())
+	}
+	ctxLogger.Debugf("Received batch request for %v", turingReqID)
+
+	ctx = h.enableTracingSpan(ctx, req)
+
+	// Read the request body
+	requestBody, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		h.error(ctx, rw, errors.NewHTTPError(err))
+		return
+	}
+
+	//Valid request
+	if valid := json.Valid(requestBody); !valid {
+		h.error(ctx, rw, errors.NewHTTPError(errors.Newf(errors.BadInput,
+			`Invalid json request format`)))
+		return
+	}
+
+	//Split into batches
+	var batchRequests map[string][]interface{}
+	err = json.Unmarshal(requestBody, &batchRequests)
+	if err != nil {
+		h.error(ctx, rw, errors.NewHTTPError(errors.Newf(errors.BadInput,
+			`Invalid json request`)))
+		return
+	}
+
+	//Validate request
+	if _, ok := batchRequests["batch_request"]; !ok {
+		h.error(ctx, rw, errors.NewHTTPError(errors.Newf(errors.BadInput,
+			`batch_request" not found in request`)))
+		return
+	}
+
+	//Handle request asynchronously
+	var batchResponses = make([]batchResponse, len(batchRequests["batch_request"]))
+	var waitGroup sync.WaitGroup
+	for index, value := range batchRequests["batch_request"] {
+		waitGroup.Add(1)
+
+		go func(index int, jsonRequestBody interface{}) {
+			defer waitGroup.Done()
+			batchRequestBody, _ := json.Marshal(jsonRequestBody)
+			var batchResponse batchResponse
+			resp, httpErr := h.getPrediction(ctx, req, ctxLogger, batchRequestBody)
+			if httpErr != nil {
+				batchResponse.StatusCode = httpErr.Code
+				batchResponse.ErrorMsg = httpErr.Message
+				batchResponses[index] = batchResponse
+				return
+			}
+			err := json.Unmarshal(resp.Body(), &batchResponse.Data)
+			if err != nil {
+				batchResponse.StatusCode = http.StatusInternalServerError
+				batchResponse.ErrorMsg = "Unable to marshall response into json"
+			}
+			batchResponse.StatusCode = http.StatusOK
+			batchResponses[index] = batchResponse
+		}(index, value)
+	}
+	waitGroup.Wait()
```

Comment:
How do we handle a timeout? Do we take into account router's timeout setting for the batch predict endpoint?

Reply:
they are handled individually in `resp, httpErr := h.getPrediction(ctx, req, ctxLogger, batchRequestBody)`

If the router timeout is 1s , individual request will get a 408 fiber timeout at 1s.
In the event of route timeout, individual request will get a 500 timeout (ss in the last section of description)

With the router timeout applied on individual requests, it will appear as though it is applied to the main request. 
I left the behaviour as such because in the event of partial success/failures, it might be more ideal to return them as it is , rather than to timeout the entire batch request. 

The only issues is that Im not sure what is the http status code to return, in the event of partial failure, keeping it as 200 for now.

Reply:
Ok, this makes sense. Resolving this thread